src/tactic/hint.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import tactic.tidy
+
+namespace tactic
+
+namespace hint
+
+/-- An attribute marking a `tactic unit` or `tactic string` which should be used by the `hint` tactic. -/
+meta def hint_attribute : user_attribute := {
+  name := `hint,
+  descr := "A tactic that should be tried by `hint`."
+}
+
+run_cmd attribute.register ``hint_attribute
+
+open lean lean.parser interactive
+
+private meta def add_tactic_hint (n : name) (t : expr) : tactic unit :=
+do
+  add_decl $ declaration.defn n [] `(tactic string) t reducibility_hints.opaque ff,
+  hint_attribute.set n () tt,
+  nolint_attr.set n () tt
+
+/--
+`add_hint t` runs the tactic `t` whenever `hint` is invoked.
+The typical use case is `add_hint "foo"` for some interactive tactic `foo`.
+-/
+@[user_command] meta def add_hint (_ : parse (tk "add_hint")) : parser unit :=
+do n ← parser.pexpr,
+   e ← to_expr n,
+   s ← eval_expr string e,
+   let t := "`[" ++ s ++ "]",
+   (t, _) ← with_input parser.pexpr t,
+   of_tactic $ do
+   let h := s ++ "_hint",
```

Comment:
```suggestion
   let h := s <.> "_hint",
```
Some tactics (and, I believe, tab complete?) will identify/ignore internal definitions if the last name component starts with `_`.

Reply:
Looks good!